### PR TITLE
Half v2plus

### PR DESCRIPTION
**Pull request by turneram**
This PR adds support for Half versions that no longer contain the `expr` type.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -18,6 +18,22 @@ else()
...
+
+include(CheckTypeSize)
+set(CMAKE_REQUIRED_INCLUDES ${HALF_INCLUDE_DIR})
+set(CMAKE_EXTRA_INCLUDE_FILES half.hpp)
+check_type_size("half_float::detail::expr" EXPR LANGUAGE CXX)
+set(CMAKE_REQUIRED_INCLUDES)
+set(CMAKE_EXTRA_INCLUDE_FILES)
+if(HAVE_EXPR)
+    add_definitions(-DHAS_HALF_V1)
+endif()
```
Comment (pfultz2) [inline CMakeLists.txt:35]: This should probably go after `find_package(ROCM REQUIRED)`

Comment (turneram) [reply CMakeLists.txt:36]: Moved

File: CMakeLists.txt [hunk 2]
```diff
@@ -19,6 +19,22 @@ endif()
...
+
+include(CheckTypeSize)
+set(CMAKE_REQUIRED_INCLUDES ${HALF_INCLUDE_DIR})
+set(CMAKE_EXTRA_INCLUDE_FILES half.hpp)
+check_type_size("half_float::detail::expr" EXPR LANGUAGE CXX)
+set(CMAKE_REQUIRED_INCLUDES)
+set(CMAKE_EXTRA_INCLUDE_FILES)
+if(HAVE_EXPR)
+    add_definitions(-DHAS_HALF_V1)
```
Comment (pfultz2) [inline CMakeLists.txt:35]: Actually, the define should be added to the migraphx target(in the src directory):

```cmake
target_compile_definitions(migraphx PUBLIC -DHAS_HALF_V1)
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -19,6 +19,22 @@ endif()
...
+    message(FATAL_ERROR "Could not find half.hpp - Please check that the install path of half.hpp has been added to CMAKE_PREFIX_PATH")
+endif()
+
+include(CheckTypeSize)
+set(CMAKE_REQUIRED_INCLUDES ${HALF_INCLUDE_DIR})
+set(CMAKE_EXTRA_INCLUDE_FILES half.hpp)
+check_type_size("half_float::detail::expr" EXPR LANGUAGE CXX)
+set(CMAKE_REQUIRED_INCLUDES)
+set(CMAKE_EXTRA_INCLUDE_FILES)
+if(HAVE_EXPR)
```
Comment (pfultz2) [inline CMakeLists.txt:34]: The variable should probably be named `HALF_HAVE_EXPR` so it wont clash with another variable accidentally.

Comment (pfultz2) [reply CMakeLists.txt:34]: Or `HAVE_HALF_EXPR`.

Comment (turneram) [reply CMakeLists.txt]: Renamed

File: CMakeLists.txt (hunk 2 above)
Comment (turneram) [reply CMakeLists.txt]: Moved/changed

**Review by pfultz2 — APPROVED** [review]

**Comment by mvermeulen** [comment]
This is one of three PRs that are all approved (others are #736 and #737) but failing one of the required Jenkin's tests.  This one and #736 fail the ONNX test and #737 fails the tidy test.

Not sure if we have flakiness going on with Jenkins but if we unstick one of these three, I'll merge it and then we can see if Jenkins unsticks for the other two as they are run.  @pfultz2 @scxiao @kahmed10

**Comment by pfultz2** [comment]
This is the error is see in jenkins:

> /src/external/hip-on-vdi/rocclr/hip_code_object.cpp:92: guarantee(false && "hipErrorNoBinaryForGpu: Coudn't find binary for current devices!")

Which usually means the machine needs to be rebooted. I'll restart the job.

**Pull request merged**